### PR TITLE
feat(headless): added recent results controller

### PR DESCRIPTION
**Pull request by nathanlb**
Added the Recent Results controller per the [design doc](https://docs.google.com/document/d/1tnQXsr2rBCnwXDI9lacdtzNQh1ziJEB-84QyVE3Ymsc/edit?usp=sharing)

### CODE REVIEW

File: packages/headless/src/controllers/recent-results-list/headless-recent-results-list.test.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+      expectContainAction(clearRecentResults);
+      expect(recentResultsList.state.results.length).toBe(0);
+      expect(
+        engine.findAsyncAction(logClearRecentResults.pending)
+      ).toBeDefined();
+    });
+
+    it('#handleRecentResultClick should validate the given index parameter', () => {
+      const validationSpy = spyOn(
```
Comment (samisayegh) [inline packages/headless/src/controllers/recent-results-list/headless-recent-results-list.test.ts:104]: Why is a validation spy needed? If we see the function call on line 110 throws, wouldn't that be enough?
With the spy, the test now depends on internal details, such as the name of the function being `validate`

File: packages/headless/src/controllers/recent-results-list/headless-recent-results-list.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+      return state.recentResults;
+    },
+
+    clear() {
+      dispatch(logClearRecentResults());
+      dispatch(clearRecentResults());
+    },
+
+    handleRecentResultClick(index: number) {
```
Comment (samisayegh) [inline packages/headless/src/controllers/recent-results-list/headless-recent-results-list.ts:151]: Could we simplify this method name to `select`? This would be more consistent with the `buildInteractiveResult` controller.

Another aspect is keeping the method names focused on what they are doing, versus suggesting how they should be used. For example, this method might be used not just on click, but also on `enter` keypress, or long-press on a touch screen device.

File: packages/headless/src/controllers/recent-results-list/headless-recent-results-list.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+      const errorMessage = new NumberValue({
+        required: true,
+        min: 0,
+        max: this.state.results.length,
+      }).validate(index);
```
Comment (samisayegh) [inline packages/headless/src/controllers/recent-results-list/headless-recent-results-list.ts:156]: The other controllers do not validate method arguments, so I think it is okay to not validate here (the dispatched action would still have validation).

On the `max` specifically, I'm not sure if there is an off-by-one since we are expecting an index, not a length?
Personally, I would place the logic in the reducer instead of the action, to avoid the complexity that comes with writing and testing thunks.

**Review by samisayegh — APPROVED** [review]
Thanks for the contribution 💯

File: packages/headless/src/features/recent-results/recent-results-actions-loader.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ */
+export function loadRecentResultsActions(
+  engine: SearchEngine
+): RecentResultsActionCreators {
+  engine.addReducers({recentResults});
+
+  return {
+    registerRecentResults,
+    clearRecentResults,
+  };
```
[the remark this answers is not in the document]
Comment (nathanlb) [reply packages/headless/src/features/recent-results/recent-results-actions-loader.ts:51]: Yes indeed

File: packages/headless/src/test/mock-result.ts [hunk 5]
```diff
@@ -3,15 +3,15 @@ import {buildMockRaw} from './mock-raw';
...
+    title: 'title',
+    uri: 'uri',
+    printableUri: 'printable-uri',
+    clickUri: 'click-uri',
+    uniqueId: 'unique-id',
+    excerpt: 'exceprt',
+    firstSentences: 'first-sentences',
     summary: null,
-    flags: '',
+    flags: 'flags',
```
[the remark this answers is not in the document]
Comment (nathanlb) [reply packages/headless/src/test/mock-result.ts]: done

File: packages/headless/src/features/recent-results/recent-results-slice.test.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+    expect(state.results).toEqual([mockResult]);
+  });
+
+  it('#pushRecentResult should add new recent result if queue is non-empty', () => {
+    state.results = [otherMockResult];
+    state = recentResultsReducer(state, pushRecentResult(mockResult));
+
+    expect(state.results.length).toBe(2);
+  });
```
Comment (jeremierobert-coveo) [inline packages/headless/src/features/recent-results/recent-results-slice.test.ts:75]: Is result order important ? If so, we should probably check element order in this test.

File: packages/headless/src/features/recent-results/recent-results-analytics-actions.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * Logs a custom event with an `actionCause` value of `clearRecentResults`.
+ */
+export const logClearRecentResults = makeAnalyticsAction(
+  'analytics/recentResults/clear',
+  AnalyticsType.Custom,
+  (client) => {
+    return client.logClearRecentResults();
+  }
+);
```
Comment (jeremierobert-coveo) [inline packages/headless/src/features/recent-results/recent-results-analytics-actions.ts:36]: ?
```suggestion
export const logClearRecentResults = makeAnalyticsAction(
  'analytics/recentResults/clear',
  AnalyticsType.Custom,
  (client) => return client.logClearRecentResults()
);
```

File: packages/headless/src/features/recent-results/recent-results-analytics-actions.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+export const logRecentResultClickThunk = (result: Result) =>
+  makeAnalyticsAction(
+    'analytics/recentResults/click',
+    AnalyticsType.Custom,
+    (client, state) => {
+      return client.logRecentResultClick(
+        partialDocumentInformation(result, state),
+        documentIdentifier(result)
+      );
+    }
```
Comment (jeremierobert-coveo) [inline packages/headless/src/features/recent-results/recent-results-analytics-actions.ts:18]:
```suggestion
    (client, state) => client.logRecentResultClick(
          partialDocumentInformation(result, state),
          documentIdentifier(result))
    }
```

File: packages/headless/src/features/recent-results/recent-results-actions.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  }),
+  maxLength: new NumberValue({required: true, min: 1, default: 10}),
+};
+
+/**
+ * Initialize the `recentResults` state.
+ * @param payload (RegisterRecentResultsCreatorPayload) The initial state and options.
+ */
+export const registerRecentResults = createAction(
+  'recentResults/registerRecentResults',
```
Comment (jeremierobert-coveo) [inline packages/headless/src/features/recent-results/recent-results-actions.ts:34]: Do you think that putting all the actions magic strings in a const object might help with readability ?

File: packages/headless/src/controllers/recent-results-list/headless-recent-results-list.ts [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+ *
+ * @param engine - The headless engine.
+ * @param props - The configuration `RecentResultsList` properties.
+ * @returns A `RecentResultsList` controller instance.
+ * */
+export function buildRecentResultsList(
+  engine: SearchEngine,
+  props?: RecentResultsListProps
+): RecentResultsList {
+  if (!loadRecentResultsListReducer(engine)) {
```
Comment (jeremierobert-coveo) [inline packages/headless/src/controllers/recent-results-list/headless-recent-results-list.ts:106]: This check is a tautology, `loadRecentResultsListReducer` can only produce true values

File: packages/headless/src/controllers/recent-results-list/headless-recent-results-list.ts [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+  dispatch(registerRecentResults(options));
+
+  return {
+    ...controller,
+
+    get state() {
+      const state = getState();
+
+      return state.recentResults;
```
Comment (jeremierobert-coveo) [inline packages/headless/src/controllers/recent-results-list/headless-recent-results-list.ts:134]:
```suggestion
      return getState().recentResults;
```

File: packages/headless/src/controllers/recent-results-list/headless-recent-results-list.ts [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+  };
+
+  validateRecentResultsProps(engine, registrationProps);
+
+  const options = {
+    results: registrationProps.initialState.results,
+    maxLength: registrationProps.options.maxLength,
+  };
+
+  dispatch(registerRecentResults(options));
```
Comment (jeremierobert-coveo) [inline packages/headless/src/controllers/recent-results-list/headless-recent-results-list.ts:126]:
```suggestion
  dispatch(registerRecentResults({
    results: registrationProps.initialState.results,
    maxLength: registrationProps.options.maxLength,
  }));
```

**Review by jeremierobert-coveo — COMMENTED** [review]
Very nice ! 🙂

File: packages/headless/src/features/recent-results/recent-results-slice.test.ts [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    expect(state.results).toEqual([mockResult, otherMockResult]);
+  });
+
+  it('#pushRecentResult should place the result at the beginning of the array if it is already present', () => {
+    state.results = [mockResult, otherMockResult];
+    state.maxLength = 2;
+    state = recentResultsReducer(state, pushRecentResult(otherMockResult));
+
+    expect(state.results).toEqual([otherMockResult, mockResult]);
+  });
```
Comment (olamothe) [inline packages/headless/src/features/recent-results/recent-results-slice.test.ts:94]: I think it's missing a test to verify that we are not adding duplicate in the list.

File: packages/headless/src/features/recent-results/recent-results-actions.ts [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+);
+
+/**
+ * Push a recent result to the list.
+ * @param payload (Result) The recently viewed result to push to the list.
+ */
+export const pushRecentResult = createAction(
+  'recentResults/pushRecentResult',
+  (payload: Result) => {
+    validateResultPayload(payload);
```
Comment (olamothe) [inline packages/headless/src/features/recent-results/recent-results-actions.ts:46]: This validator verifies properties that are needed in order to be able to properly log analytics events are present. 

It's not checking that it's a "full actual" result object, just some partial properties.

So while this seems good enough, I think this could in theory be a source of potential bugs, since we are more interested in the property that allows a result object to build back a clickable link for the end user (I assume?).

So we might instead want to have a dedicated validator just for that, that checks for a valid title and click uri (and potentially more properties down the line).

But, for the sake of making sure that a user of the library in JavaScript is passing in something that kind of ressemble a full result object (and not a string, or a number), that function can do the trick.

I let you be the judge.

**Review by olamothe — APPROVED** [review]

File: packages/headless/src/features/recent-results/recent-results-actions.ts (hunk 9 above)
Comment (nathanlb) [reply packages/headless/src/features/recent-results/recent-results-actions.ts:34]: This is the pattern that's used everywhere else in headless and I'd prefer to stay consistent. The string isn't referenced elsewhere so this magic string is only declared here.

File: packages/headless/src/features/recent-results/recent-results-actions.ts (hunk 14 above)
Comment (nathanlb) [reply packages/headless/src/features/recent-results/recent-results-actions.ts:46]: I agree. I understodd it differently as the partial schema being a "good enough" set of values to validate that it's a result. But you're right that it isn't full-proof. I'll define a complete result schema, based on the Result interface and use that.

Comment (olamothe) [reply packages/headless/src/features/recent-results/recent-results-actions.ts:46]: > 'll define a complete result schema, based on the Result interface and use that.

That might be overkill though, and by definition, it's impossible to know exactly what is in a result, because it varies depending on the context (for example, which fields are available).

So a complete validator definition will be... cumbersome, to say the least. It's going to be huge. 

Humm... 🤔 I'm wondering what are the other guys opinion. Maybe just keeping it like this is fine...

Comment (nathanlb) [reply packages/headless/src/features/recent-results/recent-results-actions.ts:46]: It would definitely be huge. I would've only bothered with the required attributes.

Comment (samisayegh) [reply packages/headless/src/features/recent-results/recent-results-actions.ts:46]: I agree, I'd keep the validation simple and focused on what's needed vs. the effort needed to be comprehensive.

File: packages/headless/src/controllers/recent-results-list/headless-recent-results-list.ts (hunk 10 above)
Comment (nathanlb) [reply packages/headless/src/controllers/recent-results-list/headless-recent-results-list.ts:106]: Every single controller does this exact thing. @samisayegh thoughts?

File: packages/headless/src/features/recent-results/recent-results-slice.ts [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+  pushRecentResult,
+} from './recent-results-actions';
+import {getRecentResultsInitialState} from './recent-results-state';
+
+export const recentResultsReducer = createReducer(
+  getRecentResultsInitialState(),
+  (builder) => {
+    builder
+      .addCase(registerRecentResults, (state, action) => {
+        state.results = action.payload.results;
```
Comment (lbergeron) [inline packages/headless/src/features/recent-results/recent-results-slice.ts:14]: We should ensure that `results` does not contain more values than the specified `maxLength`.

**Review by lbergeron — CHANGES_REQUESTED** [review]
Please validate the concerns about UA events and the results array checks.

File: packages/headless/src/features/recent-results/recent-results-slice.ts (hunk 15 above)
Comment (nathanlb) [reply packages/headless/src/features/recent-results/recent-results-slice.ts:14]: Yeah good point I can slice it to the maxLength

File: packages/headless/src/controllers/result-list/headless-interactive-result.ts [hunk 16]
```diff
@@ -90,7 +98,10 @@ export function buildInteractiveResult(
   let longPressTimer: NodeJS.Timeout;
 
   return {
-    select: logAnalyticsIfNeverOpened,
+    select: () => {
+      addToRecentResultsList();
```
[the remark this answers is not in the document]
Comment (nathanlb) [reply packages/headless/src/controllers/result-list/headless-interactive-result.ts:102]: I agree it should be added after.

I agree with your second point as well. I'm not sure yet how to make this happen. The controller could expose a `select` action for logging the analytics but then it would be on the UI to manage it and that's not ideal.
It could also be an optional prop in the `interactive-result` that signals if the result is in a that "recent-results" component. Do you have any ideas?

File: packages/headless/src/controllers/recent-results-list/headless-recent-results-list.ts (hunk 10 above)
Comment (samisayegh) [reply packages/headless/src/controllers/recent-results-list/headless-recent-results-list.ts:106]: The function always returns `true`. We use it as a type-guard to tell typescript that the engine in the `else` branch has a `RecentResultsSection`. Maybe there is a better way to do this? I prefer to avoid casts though,

File: packages/headless/src/controllers/result-list/headless-interactive-result.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (lbergeron) [reply packages/headless/src/controllers/result-list/headless-interactive-result.ts:102]: I was also thinking about using a prop in the `interactive-result`. The sad thing about it is that we'll most likely need to set some attribute on the `interactive-result` component to flag it as a "recent result". If someone forget to set it, UA events will be missing.

Introducing a new component could be an option, but we risk running into the same kind of issue we have in Salesforce (i.e., having many result links that work only in some specific situations).

I wonder whether this could be handled by the `recent-results-list` instead of the `interactive-result` itself?

[the remark this answers is not in the document]
Comment (olamothe) [reply packages/headless/src/controllers/result-list/headless-interactive-result.ts:102]: The concept of logging UA events automatically for the user works generally well in headless, * except * for clicks.

The problem with clicks is that they are heavily dependent on the end user application. A "click" can be different thing if you are inside a browser, inside a phone app, inside a desktop app, etc etc..

InteractiveResult controller is what we were able to come up with that solves some problem, but is far from perfect.

In theory this component could not exists, and user would still be able to log click event, and to decide which click event to log in which scenario.  Like, InteractiveResult controller is somewhat superfluous.

However, the main reason of it's existence is to try and "force" users to have correct analytics tracking on click, to guide them to do the right thing.

Which is why we try to promote it's usage in various code samples.

> I wonder whether this could be handled by the recent-results-list instead of the interactive-result itself?

I don't think this would work, as we can't assume any browser event there: It really has to come from the user code that hook "a click happened here, so I need to call function XYZ in headless".


If I had to choose, my instinct would be that we should have different component. To enforce the usage of the correct controller with the correct "result list", could we devise a way to better guider users that they need to use `interactive-result` with `result-list` and `interactive-recent-result` with `recent-result` ?

What if `buildInteractiveResult` function was a property returned by `buildResultList()`, and `buildInteractiveRecentResult` function was a property returned by `buildRecentResults()` ? 🤔

[the remark this answers is not in the document]
Comment (nathanlb) [reply packages/headless/src/controllers/result-list/headless-interactive-result.ts:102]: It can but then the best I have so far is my first proposition. An action exposed on the controller that logs the action. Calling it in a click handler will be the UI libs' responsibility though. It won't happen on its own in the background like most analytics.

File: packages/headless/src/features/recent-results/recent-results-actions.ts (hunk 14 above)
Comment (nathanlb) [reply packages/headless/src/features/recent-results/recent-results-actions.ts:46]: Just so we're clear, we're saying we're ok with using the current partial validation?

File: packages/headless/src/controllers/recent-results-list/headless-recent-results-list.ts (hunk 10 above)
Comment (nathanlb) [reply packages/headless/src/controllers/recent-results-list/headless-recent-results-list.ts:106]: Ah thanks it makes more sense now. I don't know if there's a better way but I'd rather not modify every other controller in this branch if we found one now 😅

File: packages/headless/src/features/recent-results/recent-results-slice.test.ts (hunk 13 above)
Comment (nathanlb) [reply packages/headless/src/features/recent-results/recent-results-slice.test.ts:94]: The test at line 88 seems to cover this. No?

File: packages/headless/src/controllers/result-list/headless-interactive-result.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (nathanlb) [reply packages/headless/src/controllers/result-list/headless-interactive-result.ts:102]: I do prefer having a separate controller for `interactive-recent-result` over having different behaviours coexist inside `interactive-result`.
Returning the `build...` function could help express what's recommended to use but in practice I think it might be a bit cumbersome to then pass that function to the child results rather than take it from CoveoHeadless directly. I'll add a `headless-interactive-recent-result` and I think that should be pretty good.

File: packages/headless/src/features/recent-results/recent-results-actions.ts (hunk 14 above)
Comment (samisayegh) [reply packages/headless/src/features/recent-results/recent-results-actions.ts:46]: If the current partial validation is validating the relevant properties, then yes 👍

File: packages/headless/src/controllers/result-list/headless-interactive-result.ts [hunk 17]
```diff
@@ -79,6 +80,11 @@ export function buildInteractiveResult(
   };
 
   let wasOpened = false;
+
+  const addToRecentResultsList = () => {
+    engine.dispatch(pushRecentResult(options.result));
```
Comment (btaillon) [inline packages/headless/src/controllers/result-list/headless-interactive-result.ts:85]: This should probably be added into `logAnalyticsIfNeverOpened`

File: packages/headless/src/controllers/recent-results-list/headless-interactive-recent-result.ts [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+
+  // 1 second is a reasonable amount of time to catch most longpress actions
+  const defaultDelay = 1000;
+  const options: Required<InteractiveRecentResultOptions> = {
+    selectionDelay: defaultDelay,
+    ...props.options,
+  };
+
+  const logAnalytics = () => {
+    engine.dispatch(logRecentResultClick(options.result));
```
Comment (btaillon) [inline packages/headless/src/controllers/recent-results-list/headless-interactive-recent-result.ts:82]: Similarly to `InteractiveResult`, this should probably only log analytics if the resutl was never opened.

File: packages/headless/src/controllers/recent-results-list/headless-interactive-recent-result.ts [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+  return {
+    select: logAnalytics,
+
+    beginDelayedSelect() {
+      longPressTimer = setTimeout(logAnalytics, options.selectionDelay);
+    },
+
+    cancelPendingSelect() {
+      longPressTimer && clearTimeout(longPressTimer);
+    },
```
Comment (btaillon) [inline packages/headless/src/controllers/recent-results-list/headless-interactive-recent-result.ts:96]: The `beginDelayedSelect` and `cancelPendingSelect` is duplicated, perhaps those could be extracted.

File: packages/headless/src/controllers/recent-results-list/headless-interactive-recent-result.ts [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+ */
+export function buildInteractiveRecentResult(
+  engine: SearchEngine,
+  props: InteractiveRecentResultProps
+): InteractiveRecentResult {
+  if (!loadInteractiveRecentResultReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  // 1 second is a reasonable amount of time to catch most longpress actions
```
Comment (samisayegh) [inline packages/headless/src/controllers/recent-results-list/headless-interactive-recent-result.ts:74]: For the implementation starting on this line down, I think we can extract and reuse the common logic between interactive result and this controller. It would also save on UT duplication, for this use-case and others in the future.

I think something like this should be possible?

```
export function buildInteractiveRecentResult(engine, props) {
    const action = () => logRecentResultClick(option.result);
    return buildCoreInteractiveResult({engine, props, action});
}
```

File: packages/headless/src/controllers/recent-results-list/headless-interactive-recent-result.ts (hunk 18 above)
Comment (nathanlb) [reply packages/headless/src/controllers/recent-results-list/headless-interactive-recent-result.ts:82]: I disagree. The result appears in the recent results list because it was opened before. And this custom event is to track usage of the recently clicked results component so we'd want to know if the result was reopened more than once.

File: packages/headless/src/controllers/result-list/headless-interactive-result.ts (hunk 17 above)
Comment (nathanlb) [reply packages/headless/src/controllers/result-list/headless-interactive-result.ts:85]: Technically this isn't an analytics action though so I preferred to have it separate. And if the result is already in the list because it was opened before then it needs to be pushed anyway because it moves it to the top of the list
